File: aircon/src/main/java/com/ironsource/aura/airconkt/AirConKt.java
package com.ironsource.aura.airconkt;

import android.app.Application;
import android.content.Context;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.ironsource.aura.airconkt.injection.AttributeResolver;
import com.ironsource.aura.airconkt.logging.Logger;
import com.ironsource.aura.airconkt.source.ConfigSource;
import com.ironsource.aura.airconkt.source.ConfigSourceRepository;

/**
 * AirCon SDK entry point.
 * Use {@link #get()} in order to get an instance.
 */
public class AirConKt {

	private static AirConKt sInstance;

	private Context mContext;

	private Logger            mLogger;
	private Class             mAttrClass;
	private AttributeResolver mAttributeResolver;
	private JsonConverter     mJsonConverter;

	private ConfigSourceRepository mConfigSourceRepository;

	/**
	 * Returns a singleton instance of the AirCon SDK.
	 * Before any use of the SDK {@link #init(AirConConfiguration)} should be called ({@link Application#onCreate()} is the common place)
	 *
	 * @return AirCon SDK instance.
	 */
	public static synchronized AirConKt get() {
		if (sInstance == null) {
			synchronized (AirConKt.class) {
				if (sInstance == null) {
					sInstance = new AirConKt();
				}
			}
		}
		return sInstance;
	}

	/**
	 * Initializes the SDK with the provided configuration.
	 *
	 * @param airConConfiguration SDK configuration.
	 */
	public void init(@NonNull AirConConfiguration airConConfiguration) {
		if (isInitialized()) {
			return;
		}
		mContext = airConConfiguration.getContext();

		mLogger = airConConfiguration.getLogger();
		mAttrClass = airConConfiguration.getAttrClass();
		mAttributeResolver = airConConfiguration.getAttributeResolver();
		mJsonConverter = airConConfiguration.getJsonConverter();

		final SdkContext sdkContext = new SdkContext(airConConfiguration.getLogger());
		mConfigSourceRepository = new ConfigSourceRepository(sdkContext, airConConfiguration.getConfigSources(), airConConfiguration.getIdentifiableConfigSources(), airConConfiguration.getIdentifiableConfigSourceFactories());
	}

	/**
	 * Returns whether the SDK was initialized.
	 *
	 * @return true is SDK is initialized, false otherwise.
	 * @see #init(AirConConfiguration)
	 */
	public boolean isInitialized() {
		return mContext != null;
	}

	/**
	 * Returns the application context retrieved from context the SDK was initialized with.
	 *
	 * @return SDK context.
	 * @throws IllegalStateException if SDK is not initialized.
	 */
	@NonNull
	public Context getContext() {
		assertInitialized();
		return mContext;
	}

	/**
	 * Returns the SDK logger set by {@link AirConConfiguration.Builder#setLogger(Logger)}
	 * or {@link com.ironsource.aura.airconkt.logging.AndroidLogger} if non is provided.
	 *
	 * @return SDK logger.
	 * @throws IllegalStateException if SDK is not initialized.
	 */
	@NonNull
	public Logger getLogger() {
		assertInitialized();
		return mLogger;
	}

	/**
	 * Returns whether XML injection is enabled.
	 *
	 * @return true if XML injection is enabled, false otherwise.
	 * @throws IllegalStateException if SDK is not initialized.
	 * @see AirConConfiguration.Builder#enableXmlInjection(Class, AttributeResolver)
	 */
	public boolean isXmlInjectionEnabled() {
		assertInitialized();
		return mAttrClass != null;
	}

	/**
	 * Returns the attribute class if supplied, null otherwise.
	 *
	 * @return attribute class if supplied, null otherwise.
	 * @throws IllegalStateException if SDK is not initialized.
	 * @see AirConConfiguration.Builder#enableXmlInjection(Class, AttributeResolver)
	 */
	@Nullable
	public Class getAttrClass() {
		assertInitialized();
		return mAttrClass;
	}

	/**
	 * Returns attribute resolver if supplied, null otherwise.
	 *
	 * @return attribute resolver if supplied, null otherwise.
	 * @throws IllegalStateException if SDK is not initialized.
	 * @see AirConConfiguration.Builder#enableXmlInjection(Class, AttributeResolver)
	 */
	public AttributeResolver getAttributeResolver() {
		assertInitialized();
		return mAttributeResolver;
	}

	/**
	 * Returns json converter if supplied..
	 *
	 * @return json converter if supplied, null otherwise.
	 * @throws IllegalStateException if SDK is not initialized or no converter is supplied.
	 * @see AirConConfiguration.Builder#setJsonConverter(JsonConverter)
	 */
	public JsonConverter getJsonConverter() {
		assertInitialized();

		if (mJsonConverter == null) {
			throw new IllegalStateException("No json converter available, a converter needs to be supplied in the AirConConfiguration");
		}

		return mJsonConverter;
	}

	/**
	 * Returns config sources repository through which {@link ConfigSource}
	 * can be added, removed and retrieved.
	 *
	 * @return config source repository.
	 * @throws IllegalStateException if SDK is not initialized.
	 */
	@NonNull
	public ConfigSourceRepository getConfigSourceRepository() {
		assertInitialized();
		return mConfigSourceRepository;
	}

	private void assertInitialized() {
		if (!isInitialized()) {
			throw new IllegalStateException("Sdk not initialized, did you call AirCon.init()?");
		}
	}
}

File: aircon/src/main/java/com/ironsource/aura/airconkt/AirConKt.kt
package com.ironsource.aura.airconkt

import android.content.Context
import com.ironsource.aura.airconkt.config.type.util.JsonConverter
import com.ironsource.aura.airconkt.logging.AndroidLogger
import com.ironsource.aura.airconkt.logging.Logger
import com.ironsource.aura.airconkt.source.ConfigSource
import com.ironsource.aura.airconkt.source.ConfigSourceRepository
import com.ironsource.aura.dslint.annotations.DSLint

/**
 * AirCon SDK entry point.
 */
object AirConKt {

    /**
     * Initializes the SDK with configuration
     */
    fun init(context: Context,
             block: Options.() -> Unit) {
        if (initialized) {
            return
        }

        this.context = context.applicationContext

        val options = OptionsBuilder(block)

        if (options.loggingOptions.enabled) {
            logger = options.loggingOptions.logger
        }

        if (options.hasJsonConverter()) {
            jsonConverter = options.jsonConverter
        }

        configSourceRepository = options.configSourceRepository
    }

    private var initialized = ::context.isInitialized

    lateinit var context: Context
        private set

    var logger: Logger? = null
        private set

    var jsonConverter: JsonConverter? = null
        get() {
            checkNotNull(
                    field) { "No json converter available, a converter needs to be supplied in AirCon.init()" }
            return field
        }
        private set

    /**
     * Returns config sources repository through which [ConfigSource]
     * can be added, removed and retrieved.
     *
     * @return config source repository.
     * @throws IllegalStateException if SDK is not initialized.
     */
    lateinit var configSourceRepository: ConfigSourceRepository
        private set
}

/**
 * Configuration class used to initialize the SDK.
 *
 * @see AirConKt.init
 */
@DSLint
interface Options {

    /**
     * Json converter to be used with jsonConfig
     */
    var jsonConverter: JsonConverter

    /**
     * Define SDK logging options
     */
    fun logging(block: LoggingOptions.() -> Unit)

    /**
     * Add a config source.
     * A config can have only one instance of the same class.
     */
    fun configSource(configSource: () -> ConfigSource)
}

private class OptionsBuilder : Options {

    companion object {
        operator fun invoke(block: Options.() -> Unit) = OptionsBuilder().apply(
                block)
    }

    override lateinit var jsonConverter: JsonConverter

    internal val configSourceRepository = ConfigSourceRepository()

    internal fun hasJsonConverter() = ::jsonConverter.isInitialized

    internal var loggingOptions: LoggingOptions = LoggingOptionsBuilder {}

    override fun logging(block: LoggingOptions.() -> Unit) {
        loggingOptions = LoggingOptionsBuilder(block)
    }

    override fun configSource(configSource: () -> ConfigSource) {
        configSourceRepository.addSource(configSource())
    }
}

@DSLint
interface LoggingOptions {

    /**
     * Set whether SDK logging is enabled (true by default).
     */
    var enabled: Boolean

    /**
     * Set a logger to be used by the SDK.
     * If a logger is not supplied [com.ironsource.aura.airconkt.logging.AndroidLogger] is used.
     */
    var logger: Logger
}

private class LoggingOptionsBuilder : LoggingOptions {

    companion object {
        operator fun invoke(block: LoggingOptionsBuilder.() -> Unit) = LoggingOptionsBuilder().apply(
                block)
    }

    /**
     * Set whether SDK logging is enabled (true by default).
     */
    override var enabled = true

    /**
     * Set a logger to be used by the SDK.
     * If a logger is not supplied [com.ironsource.aura.airconkt.logging.AndroidLogger] is used.
     */
    override var logger: Logger = AndroidLogger()
}